GoogleDataTransport/GDTCCTLibrary/GDTCCTUploadOperation.m

```diff
@@ -0,0 +1,529 @@
+  GDTCORClock *nextUploadTime = [self.metadataProvider nextUploadTimeForTarget:target];
+  if (nextUploadTime) {
+    isAfterNextUploadTime = [[GDTCORClock snapshot] isAfter:nextUploadTime];
+  }
+
+  if (isAfterNextUploadTime) {
+    GDTCORLogDebug(@"CCT: can upload to target %ld because the request wait time has transpired",
+                   (long)target);
+  } else {
+    GDTCORLogDebug(@"CCT: can't upload to target %ld because the backend asked to wait",
+                   (long)target);
+  }
+
+  return isAfterNextUploadTime;
+}
+
+/** Constructs data given an upload package.
+ *
+ * @param events The events used to construct the request proto bytes.
+ * @return Proto bytes representing a gdt_cct_LogRequest object.
+ */
+- (nonnull NSData *)constructRequestProtoWithEvents:(NSSet<GDTCOREvent *> *)events {
+  // Segment the log events by log type.
+  NSMutableDictionary<NSString *, NSMutableSet<GDTCOREvent *> *> *logMappingIDToLogSet =
+      [[NSMutableDictionary alloc] init];
+  [events enumerateObjectsUsingBlock:^(GDTCOREvent *_Nonnull event, BOOL *_Nonnull stop) {
+    NSMutableSet *logSet = logMappingIDToLogSet[event.mappingID];
+    logSet = logSet ? logSet : [[NSMutableSet alloc] init];
+    [logSet addObject:event];
+    logMappingIDToLogSet[event.mappingID] = logSet;
+  }];
+
+  gdt_cct_BatchedLogRequest batchedLogRequest =
+      GDTCCTConstructBatchedLogRequest(logMappingIDToLogSet);
+
+  NSData *data = GDTCCTEncodeBatchedLogRequest(&batchedLogRequest);
+  pb_release(gdt_cct_BatchedLogRequest_fields, &batchedLogRequest);
+  return data ? data : [[NSData alloc] init];
+}
+
+/** Constructs a request to FLL given a URL and request body data.
+ *
+ * @param target The target backend to send the request to.
+ * @param data The request body data.
+ * @return A new NSURLRequest ready to be sent to FLL.
+ */
+- (nullable NSURLRequest *)constructRequestWithURL:(NSURL *)URL
+                                         forTarget:(GDTCORTarget)target
+                                              data:(NSData *)data {
+  if (data == nil || data.length == 0) {
+    GDTCORLogDebug(@"There was no data to construct a request for target %ld.", (long)target);
+    return nil;
+  }
+
+  NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:URL];
+  NSString *targetString;
+  switch (target) {
+    case kGDTCORTargetCCT:
+      targetString = @"cct";
+      break;
+
+    case kGDTCORTargetFLL:
+      targetString = @"fll";
+      break;
+
+    case kGDTCORTargetCSH:
+      targetString = @"csh";
+      break;
+    case kGDTCORTargetINT:
+      targetString = @"int";
+      break;
+
+    default:
+      targetString = @"unknown";
+      break;
+  }
+  NSString *userAgent =
+      [NSString stringWithFormat:@"datatransport/%@ %@support/%@ apple/", kGDTCORVersion,
+                                 targetString, kGDTCCTSupportSDKVersion];
+
+  [request setValue:[self.metadataProvider APIKeyForTarget:target]
```

Comment:
I suppose there's no particular harm for including this backends that don't support it, except the additional uploaded data, but consider making it FLL only

Reply:
I think we are actually fine here. The method `APIKeyForTarget:` returns `nil` for targets that don't require an API key. If `nil` is passed to the `setValue:forHTTPHeaderField:` method than the header is ignored (or removed if previously set).  Thank you for checking!